Update documentation on plugin installation and management

fmoralesc opened the pull request:
When working on #3125 I noticed the plugin installation and management in usr_05.txt was a bit outdated and kinda hacked up, so I tried to improve it.

fmoralesc committed:
Update documentation on plugin installation and management

Helped-by: @brcolow
Helped-by: @ZyX-I

ZyX-I commented on runtime/doc/usr_05.txt:
```diff
@@ -250,61 +250,87 @@ The ":map" command (with no arguments) lists your current mappings.  At
-
-GLOBAL PLUGINS						*standard-plugin*
-
+*05.4*	Plugins					*add-plugin* *plugin*
```
When you rename a section of the user manual you need to rename it also at the top of the file and in `usr_toc.txt`.

fmoralesc replied:
:+1:

ZyX-I commented on runtime/doc/usr_05.txt:
```diff
@@ -250,61 +250,87 @@ The ":map" command (with no arguments) lists your current mappings.  At
+    https://github.com/search?q=language%3AVimL&type=Repositories
 - They are sometimes posted in a Vim |maillist|.
-- You could write one yourself, see |write-plugin|.
+- You could write one yourself ;) (see |write-plugin|).
```
Does not smiley end a sentense? I would write the ending as `See |write-plugin|.`

ZyX-I commented on runtime/doc/usr_05.txt:
```diff
@@ -250,61 +250,87 @@ The ":map" command (with no arguments) lists your current mappings.  At
-- You could write one yourself, see |write-plugin|.
+- You could write one yourself ;) (see |write-plugin|).
+
+
```
Previous section was prefixed by a single newline.

ZyX-I commented on runtime/doc/usr_05.txt:
```diff
@@ -250,61 +250,87 @@ The ":map" command (with no arguments) lists your current mappings.  At
+have a plugin directory yet): >
 
 	mkdir ~/.vim
 	mkdir ~/.vim/plugin
```
This is ~/.nvim.

ZyX-I commented on runtime/doc/usr_05.txt:
```diff
@@ -313,38 +339,12 @@ Example for Unix (assuming you didn't have a plugin directory yet): >
-As an example, consider using "~/.vim/plugin/perl/*.vim" for all your Perl
+Instead of putting plugins directly into the plugin/ directory, you may better
+organize them by putting them into subdirectories under plugin/.  As an
+example, consider using "~/.vim/plugin/git/*.vim" for all the Git-related
```
~/.nvim.

ZyX-I commented on runtime/doc/usr_05.txt:
```diff
@@ -387,8 +383,91 @@ Only files of this filetype will use the settings from the plugin.  The <name>
+once you install a few plugins, specially if they include more than one file
+(which is common). A way to solve this is to install plugins in their own
+directories and then add the path of those to the 'runtimepath' setting (so
+Vim will use the files from those directories when looking for plugins). In
```
Two spaces after a dot.

ZyX-I commented on runtime/doc/usr_05.txt:
```diff
@@ -387,8 +383,91 @@ Only files of this filetype will use the settings from the plugin.  The <name>
+Vim will use the files from those directories when looking for plugins). In
+practice, this means that instead of having a |vimfiles| tree like >
+
+	~.vim/
```
`~/.nvim`, not `~.vim`.

ZyX-I commented on runtime/doc/usr_05.txt:
```diff
@@ -387,8 +383,91 @@ Only files of this filetype will use the settings from the plugin.  The <name>
+
+you can have something like this >
+
+	~.vim/
```
`~/.nvim`

ZyX-I commented on runtime/doc/usr_05.txt:
```diff
@@ -387,8 +383,91 @@ Only files of this filetype will use the settings from the plugin.  The <name>
+			e.vim
+
+This method allows the plugins to be distributed as their own trees, so they
+can be installed without affecting other plugins. This is very helpful when
```
Two spaces after dot.

ZyX-I commented on runtime/doc/usr_05.txt:
```diff
@@ -387,8 +383,91 @@ Only files of this filetype will use the settings from the plugin.  The <name>
+is no "bundles" directory)
+
+    1. creating the plugin directory >
+	$ mkdir -p ~/.vim/bundles/awesome-vim
```
~/.nvim

ZyX-I commented on runtime/doc/usr_05.txt:
```diff
@@ -387,8 +383,91 @@ Only files of this filetype will use the settings from the plugin.  The <name>
+    1. creating the plugin directory >
+	$ mkdir -p ~/.vim/bundles/awesome-vim
+<    2. copying the files into the "bundles/awesome-vim/" folder >
+	$ cp awesome-vim/* ~/.vim/bundles/awesome-vim/
```
~/.nvim

ZyX-I commented on runtime/doc/usr_05.txt:
```diff
@@ -387,8 +383,91 @@ Only files of this filetype will use the settings from the plugin.  The <name>
+	$ cp awesome-vim/* ~/.vim/bundles/awesome-vim/
+<    3. adding >
+
+	set runtimepath+=~/.vim/bundles/awesome-vim
```
~/.nvim. Also unlike other code blocks this one contains empty lines before and after which is strange.

fmoralesc replied:
Yeah, I thought it looked weird too close to the surrounding lines since this is not prefixed by a `$`, like in the other cases. It looks clearer this way to me. I could give more room to the previous codeblocks, but I won't remove the empty lines in this one.

ZyX-I replied:
29.09.2015, 01:36, "Felipe Morales" <notifications@github.com>:
> In runtime/doc/usr_05.txt:
>
>> + awesome.vim > + plugin/ > + awesome.vim > + doc/ > + awesome.txt > + > +from http://github.com/developer/awesome-vim, they can do so by (assuming there > +is no "bundles" directory) > + > + 1. creating the plugin directory > > + $ mkdir -p ~/.vim/bundles/awesome-vim > +< 2. copying the files into the "bundles/awesome-vim/" folder > > + $ cp awesome-vim/* ~/.vim/bundles/awesome-vim/ > +< 3. adding > > + > + set runtimepath+=~/.vim/bundles/awesome-vim
>
> Yeah, I thought it looked weird too close to the surrounding lines since this is not prefixed by a $, like in the other cases. It looks clearer this way to me. I could give more room to the previous codeblocks, but I won't remove the empty lines in this one.

It does not matter as long as you are consistent. If you think that less space here looks bad, add more space there.

>
> —
> Reply to this email directly or view it on GitHub.

ZyX-I commented on runtime/doc/usr_05.txt:
```diff
@@ -387,8 +383,91 @@ Only files of this filetype will use the settings from the plugin.  The <name>
+
+<       to their |vimrc|.
+
+Most commonly, instead of steps 1-2, this should work: >
```
Not sure, but should not it worded as “Most common”?

fmoralesc replied:
"commonly" is working as an adverb here, I think "most common" wouldn't be right.

ZyX-I commented on runtime/doc/usr_05.txt:
```diff
@@ -387,8 +383,91 @@ Only files of this filetype will use the settings from the plugin.  The <name>
+
+Please refer to their documentation on how to use them.
+
+
```
One empty line.

ZyX-I commented on runtime/doc/usr_05.txt:
```diff
@@ -387,8 +383,91 @@ Only files of this filetype will use the settings from the plugin.  The <name>
+
+Most commonly, instead of steps 1-2, this should work: >
+
+    $ mkdir ~/.vim/bundles/
```
~/.nvim

ZyX-I commented on runtime/doc/usr_05.txt:
```diff
@@ -387,8 +383,91 @@ Only files of this filetype will use the settings from the plugin.  The <name>
+Most commonly, instead of steps 1-2, this should work: >
+
+    $ mkdir ~/.vim/bundles/
+    $ cd ~/.vim/bundles
```
~/.nvim

ZyX-I commented on runtime/doc/usr_05.txt:
```diff
@@ -387,8 +383,91 @@ Only files of this filetype will use the settings from the plugin.  The <name>
+
+PLUGIN MANAGERS                                            *plugin-managers*
+
+Bundle-style plugins can be handled by so-called 'plugin managers', which are
```
`"plugin manager"`. `''` is for options

ZyX-I commented on runtime/doc/usr_05.txt:
```diff
@@ -250,61 +250,87 @@ The ":map" command (with no arguments) lists your current mappings.  At
+
+                                                        *ftplugins*
+Filetype plugins ("ftplugins") are only used for a specific type of file, and
+usually change the buffer-local settings. For example, the standard vim C
```
Two spaces after a dot.

ZyX-I commented on runtime/doc/usr_05.txt:
```diff
@@ -250,61 +250,87 @@ The ":map" command (with no arguments) lists your current mappings.  At
+Filetype plugins ("ftplugins") are only used for a specific type of file, and
+usually change the buffer-local settings. For example, the standard vim C
+ftplugin sets 'formatoptions' to break comment lines but not other lines, and
+to insert the comment leader when hitting <CR> or using |o|. Filetype plugins
```
Two spaces.

ZyX-I commented on runtime/doc/usr_05.txt:
```diff
@@ -250,61 +250,87 @@ The ":map" command (with no arguments) lists your current mappings.  At
+There are mainly two kinds of plugin: global and filetype-specific.
+
+Global plugins are sourced on startup and usually add functionality that will
+be available regardless of the opened files. For example, the |matchparen|
```
Two spaces.

ZyX-I commented on runtime/doc/usr_05.txt:
```diff
@@ -378,17 +373,102 @@ The generic names for the filetype plugins are: >
+
+<       to their |vimrc|.
+
+Most commonly, instead of steps 1-2, this should work: >
```
Adverb which modifies what? In this position with these commas this looks like an introductory phrase (where adverb has nothing to modify because this is rather independent) and not something that modifies verb “work”. I would rephrase this as “Most likely steps 1 and 2 can be replaced with” because I am pretty sure “instead of steps 1-2” could not be placed in that position.

fmoralesc replied:
Sure.

ZyX-I commented:
LGTM

justinmk commented:
I liked the brevity of the old copy. Also not sure about some of the additional tags. Let's set this for 0.2.

brcolow commented on runtime/doc/usr_05.txt:
```diff
@@ -250,121 +250,115 @@ The ":map" command (with no arguments) lists your current mappings.  At
+Vim's functionality can be extended by adding plugins.  A plugin is
+essentially a set of Vim script files that are loaded automatically when Vim
+starts or on some other event (like opening a file of a certain type).
+Plugins can be used to add support to more programming languages, extend Vim's
```
Should read: 
Plugins can be used to add support *for* more programming languages,

brcolow commented on runtime/doc/usr_05.txt:
```diff
@@ -250,121 +250,115 @@ The ":map" command (with no arguments) lists your current mappings.  At
+Plugins can be used to add support to more programming languages, extend Vim's
+command "vocabulary", simplify some use-cases, add some bling to Vim, and even
+turn it into a different sort of application.  Thankfully, the community has
+made available on the net a great ammount of plugins that any user can benefit
```
Maybe this should read:

"Thankfully there is a vibrant and active community of plugin authors from around the globe that every user can benefit from."

brcolow commented on runtime/doc/usr_05.txt:
```diff
@@ -250,121 +250,115 @@ The ":map" command (with no arguments) lists your current mappings.  At
+from.
+
+You can add a plugin very easily by dropping it in your |vimfiles| directory,
+but there are more advanced ways to manage plugins, which will be described
```
Instead of advanced, maybe convenient would be a better fit.

brcolow commented on runtime/doc/usr_05.txt:
```diff
@@ -250,121 +250,115 @@ The ":map" command (with no arguments) lists your current mappings.  At
+
+KINDS OF PLUGIN                                         *plugin-types*
+
+There are mainly two kinds of plugin: global and filetype-specific.
```
Should read: There are mainly two kinds of plugin*s*:....

brcolow commented on runtime/doc/usr_05.txt:
```diff
@@ -250,121 +250,115 @@ The ":map" command (with no arguments) lists your current mappings.  At
+There are mainly two kinds of plugin: global and filetype-specific.
+
+Global plugins are sourced on startup and usually add functionality that will
+be available regardless of the opened files.  For example, the |matchparen|
```
Maybe should read:

Global plugins are sourced on startup and usually add functionality that is available regardless of which files are open.

brcolow commented on runtime/doc/usr_05.txt:
```diff
@@ -250,121 +250,115 @@ The ":map" command (with no arguments) lists your current mappings.  At
+usually change the buffer-local settings.  For example, the standard vim C
+ftplugin sets 'formatoptions' to break comment lines but not other lines, and
+to insert the comment leader when hitting <CR> or using |o|.  Filetype plugins
+are usually distributed alongside syntax files, and indent and filetype
```
I don't like the double and here, as in "...usually distributed alongside syntax files, *and* indent *and* filetype detection scripts..." but I am not sure how to fix it.

brcolow commented on runtime/doc/usr_05.txt:
```diff
@@ -250,121 +250,115 @@ The ":map" command (with no arguments) lists your current mappings.  At
+Filetype plugins reside under the `ftplugin/` |vimfiles| directory.
+
+The Vim distribution comes with a set of plugins for different filetypes that
+you can start using with this command: >
```
Not sure about this one, but maybe it should read:

you can start using with the '>' command

(not sure about single quotes format)

brcolow commented on runtime/doc/usr_05.txt:
```diff
@@ -250,121 +250,115 @@ The ":map" command (with no arguments) lists your current mappings.  At
 - Some come with Vim.  You can find them in the directory $VIMRUNTIME/macros
   and its sub-directories.
-- Download from the net.  There is a large collection on http://www.vim.org.
+- In the net:
```
Should read:

On the net:

brcolow commented on runtime/doc/usr_05.txt:
```diff
@@ -250,121 +250,115 @@ The ":map" command (with no arguments) lists your current mappings.  At
 Some plugins can be updated automatically, see |getscript|.
 
+For all other cases, you should be able to follow the guidelines below (also,
+it's very likely that the plugin creators has given instructions on how to
```
Should read:

it's very likely that the plugins creators *have* given instructions on how...

brcolow commented on runtime/doc/usr_05.txt:
```diff
@@ -250,121 +250,115 @@ The ":map" command (with no arguments) lists your current mappings.  At
-	Unix		~/.vim/plugin/
-	Macintosh	$VIM:vimfiles:plugin
-	Mac OS X	~/.vim/plugin/
+For example, to install the justify.vim plugin in *nix (assuming you didn't
```
in *nix -> on *nix

brcolow commented on runtime/doc/usr_05.txt:
```diff
@@ -378,17 +372,102 @@ The generic names for the filetype plugins are: >
+PLUGIN MANAGERS                                            *plugin-managers*
+
+Bundle-style plugins can be handled by so-called "plugin managers", which are
+special plugins whose purpose is to handle 'runtimepath' and help the user to
```
Instead of "special plugins whose purpose is to..." maybe "special purpose plugins that handle 'runtimepath'..."

brcolow commented on runtime/doc/usr_05.txt:
```diff
@@ -378,17 +372,102 @@ The generic names for the filetype plugins are: >
+special plugins whose purpose is to handle 'runtimepath' and help the user to
+install, update and load plugins.
+
+The most popular plugin managers are:
```
Instead of "The most popular", for reasons of maintenance and future-proofing, maybe we should say "Some great plugin managers are" or "Recommended plugin managers are" or something like that.

fmoralesc commented:
Thanks, @brcolow, I'll include most (if not all) of your suggestions.

Commit:
option: fix off-by-one error when handling &directory

a8e18d9 introduced an off-by-one error that caused the last character of
&directory to be ignored. This commit is a straightforward fix for that error.

fixes #3519

jrobeson commented:
doesn't this require changes due to the XDG PR being merged?

fmoralesc commented:
@jrobeson Sure does, thanks for the reminder.

justinmk commented:
Might need to revisit this in light of and friends.

justinmk commented:
I tried to merge this, but there's too much that's out of date. And the section on "bundles" is now replaced by Vim 8 "packages".
